pkg/controller/migmigration/task.go

```diff
@@ -27,6 +29,95 @@ const (
 	Completed               = "Completed"
 )
 
+var PhaseOrder = map[string]int{
+	Created:                 00, // 0-499 normal
+	Started:                 1,
+	WaitOnResticRestart:     10,
+	ResticRestartCompleted:  11,
+	BackupStarted:           20,
+	BackupCompleted:         21,
+	WaitOnBackupReplication: 30,
+	BackupReplicated:        31,
+	RestoreStarted:          40,
+	RestoreCompleted:        41,
+	BackupFailed:            501, // 500-999 errors
+	RestoreFailed:           510,
+	Completed:               1000, // Succeeded
+}
+
+// Phase
+type Phase struct {
+	Name string
+}
+
+// Validate the phase.
+func (p Phase) Validate() error {
+	_, found := PhaseOrder[p.Name]
+	if !found {
+		m := fmt.Sprintf("phase not valid: %s", p.Name)
+		return errors.New(m)
+	}
+
+	return nil
+}
+
+// Set the phase.
+// Prevents rewind.
+func (p *Phase) Set(name string) {
+	if p.Before(name) {
+		p.Name = name
+	}
+}
+
+// Phase is equal.
+func (p Phase) Equals(other string) bool {
+	return p.Name == other
+}
+
+// Phase is after `other`.
+func (p Phase) After(other string) bool {
+	nA, found := PhaseOrder[p.Name]
+	if !found {
+		nA = 0
+	}
+	nB, found := PhaseOrder[other]
+	if !found {
+		nB = 0
```

Comment:
Handling of `!found` values seems a little weird here, since they're compared as if they're valued at zero. InvalidPhase.after(validPhase) will always be false, and InvalidPhase.before(validPhase) will always be true. Reverse the results for validPhase.after(invalidPhase), and both invalid will produce false for both before and after. Maybe this doesn't matter so much since we're already validating phase at the beginning of the task, though.

Reply:
Dealing with `!found` is difficult to deal with.  However, I don't think this is worse than how it currently works (or would work with _int_ phases).  If either the task.Phase or the value used in the comparison isn't one of the _const_ values, the result will be unpredictable.  Would be nice if Go supported enums so we should have the compiler catch it.